sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java

```diff
@@ -658,21 +657,31 @@ private String parentResourceLinkToQueryLink(String parentResouceLink, ResourceT
                 retryPolicyInstance.onBeforeSendRequest(request);
             }
 
-            return this.delete(request).map(response -> toResourceResponse(response, DocumentCollection.class));
+            return this.delete(request, retryPolicyInstance).map(response -> toResourceResponse(response, DocumentCollection.class));
 
         } catch (Exception e) {
             logger.debug("Failure in deleting a collection, due to [{}]", e.getMessage(), e);
             return Flux.error(e);
         }
     }
 
-    private Flux<RxDocumentServiceResponse> delete(RxDocumentServiceRequest request) {
+    private Flux<RxDocumentServiceResponse> delete(RxDocumentServiceRequest request, DocumentClientRetryPolicy documentClientRetryPolicy) {
         populateHeaders(request, HttpConstants.HttpMethods.DELETE);
+        if(request.requestContext != null && documentClientRetryPolicy.getRetryCount() > 0) {
+            documentClientRetryPolicy.updateEndTime();
+            request.requestContext.updateRetryContext(documentClientRetryPolicy);
+        }
+
         return getStoreProxy(request).processMessage(request);
     }
 
-    private Flux<RxDocumentServiceResponse> read(RxDocumentServiceRequest request) {
+    private Flux<RxDocumentServiceResponse> read(RxDocumentServiceRequest request, DocumentClientRetryPolicy documentClientRetryPolicy) {
         populateHeaders(request, HttpConstants.HttpMethods.GET);
+        if(request.requestContext != null && documentClientRetryPolicy.getRetryCount() > 0) {
+            documentClientRetryPolicy.updateEndTime();
+            request.requestContext.updateRetryContext(documentClientRetryPolicy);
```

Comment:
same question which I had raised offline.

I see that you have different way of capturing retry info in the direct mode, however here on the outer surface of the SDK we are again capturing the retry info, why do we need to capture the retry info in two different places? 

is this only applicable to GW mode if so it should be in GatewayStoreModel. no?

same for all other HttpMethods

Reply:
Yes above is applicable to GW mode , and retry in gateway happens via BackoffRetry.executeRetry and at that place we dont have the access to DocumentServiceRequestContext

Reply:
Discussed offline - 
"Lets take the example of ReadDocument()
In Gateway mode ReadDocument() will keep calling readDocumentInternal() based on BackOffRetryUtility.executeRetry() ,and each retry will call the readDocumentInternal() again and update
request.requestContext.updateRetryContext(documentClientRetryPolicy) everytime.
However in Direct mode  ReadDocument() will call readDocumentInternal(), and in case of any  direct exception (Gone) instead of call readDocumentInternal() via  BackOffRetryUtility.executeRetry(), replicate resource client will create it own retry policy and consume failure/success and response will eventually be send via StoreClient to ReadDocument , therefore the code
request.requestContext.updateRetryContext(documentClientRetryPolicy) in read() in RxDocumentClientImpl will never be executed again."